### ISSUE TITLE
Allow user code to stop a pipeline on demand

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#3821]
The ability to stop a pipeline on demand is currently only available _internally_, as used by `Select-Object -First <n>`, for instance.

To quote from [this uservoice.com issue](https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/11087865-enable-users-to-stop-pipeline-making-stopupstreamc):

> It is of great value to be able to stop the pipeline if your mission is completed before the emitting cmdlet has provided all results.
>
> In fact, the PS pipeline would gain a completely new "feature". Currently, it is a memory throttle and provides realtime feedback at the cost of performance. If stopping the pipeline was something available to users, it could also be a generic means of running cmdlets only partially.
> 
> Select-Object is a great start but you do not always know beforehand how many elements you need. So it would be beneficial to have a `Stop-Pipeline` cmdlet or at least a public `StopUpstreamCommandsException` that a script programme can throw.

```powershell
PowerShell Core v6.0.0-beta (v6.0.0-beta.1)
```

**Comment by dragonwolf83** [comment]
There is a good detailed blog post on how to implement this in PowerShell today: [Cancelling a Pipeline](http://community.idera.com/powershell/powershell_com_featured_blogs/b/tobias/posts/cancelling-a-pipeline). I think it would be useful to implement something like as a built-in cmdlet.

**Comment by mklement0** [comment]
Thanks, @dragonwolf83; the blog post's author, Tobias Weltner, is the same person who created the linked uservoice.com issue; the blog post elicited the following @jpsnover  response (dated "over 7 years ago"):

> We have wanted to add pipeline stopping for quite a while and it never got raised above the bar.  

That's one lofty bar! :)

**Comment by SteveL-MSFT** [comment]
Things that don't necessarily align with the PowerShell Team's priorities (which is not static) can still get done if contributed by the community :)

**Comment by iSazonov** [comment]
We need RFC for new cmdlet and examples of how to use the cmdlet.

**Comment by alx9r** [comment]
Does anyone have a working link to Tobias' blog post mentioned above?  That link seems to be dead.

**Comment by TobiasPSP** [comment]
Here‘s some material: http://www.powertheshell.com/stop_pipeline/

**Comment by SeeminglyScience** [comment]
Adding `Needs-Triage` so `WG-Engine` can discuss it. This comes up surprisingly often. Typically for profile functions, but still. The lengths one has to go through to get the same behavior is pretty wild (and unsupported).

My proposal:

```diff
  namespace System.Management.Automation
  {
      public abstract class PSCmdlet
      {
+         public void StopUpstreamCommands();
      }
  }
```

If added this should be tested both for binary cmdlets (`this.StopUpstreamCommands()`) and script cmdlets (`$PSCmdlet.StopUpstreamCommands()`)

**Comment by vexx32** [comment]
WG-Engine discussed this today (6/23/2022).

We agree that users should have access to this functionality in some way that is not exclusive to `Select-Object`. It came up in discussion that there may be undesirable elements of the current behaviour of `Select-Object -First $x`; for instance, it doesn't appear to call upstream C# cmdlets' `StopProcessing()` when it cancels the remainder of a pipeline.

We need more investigation on whether this behaviour would be problematic to expose as a commonly usable API, and whether we should:

- Make changes to the existing behaviour of the pipeline's handling of `StopUpstreamCommandsException`, or
- Expose the current behaviour as-is as a public API similar to the suggestions in this issue

**Comment by SeeminglyScience** [comment]
I went through every command that overrides `StopProcessing` but does not implement `IDisposable` with this command:

```powershell
Find-Type -Base System.Management.Automation.PSCmdlet |
    Find-Type -Not -ImplementsInterface System.IDisposable |
    Find-Member StopProcessing -Force -Declared |
    ForEach-Object {
        dnSpy.Console.exe $_.Module.Assembly.Location --md $_.MetadataToken --no-tokens |
            bat --language cs --style 'grid,numbers,snip'

        Read-Host -Prompt 'Press enter to continue'
    }
```

Of all of them I found only one command that is using `StopProcessing` for resource cleanup (`Get-Counter`).

I would recommend we just fix `Get-Counter` as pipeline terminating errors, `Select-Object -First 1` and pipeline stop exceptions thrown outside of stop requests will already cause a leak here.

**Comment by TobiasPSP** [comment]
Here is a prototype taking advantage of  the existing behavior in `Select-Object`:

````powershell
function Stop-Pipeline
{
  $wrappedCmd = $ExecutionContext.InvokeCommand.GetCommand('Microsoft.PowerShell.Utility\Select-Object', [System.Management.Automation.CommandTypes]::Cmdlet)
  $scriptCmd = {& $wrappedCmd -First 1 }
  $steppablePipeline = $scriptCmd.GetSteppablePipeline($myInvocation.CommandOrigin)
  $steppablePipeline.Begin($true)
  $steppablePipeline.Process(1)
  $steppablePipeline.End()
  $steppablePipeline.Dispose()
}

"Start"
1..10  | ForEach-Object {  if ($_ -eq 5) { Stop-Pipeline }; $_ }
"End"
````

Just make sure `Stop-Pipeline` cannot be called outside a pipeline, or else you get the same behavior as with `break`: it stops your entire script. The exception raised internally by `Select-Object` must be caught inside a pipeline and not be allowed to escape into the script scope.

**Comment by SeeminglyScience** [comment]
There's a few caveats to that though. Unfortunately I don't remember most of them, but an example is if you switch from `ForEach-Object` to `& { process {` it starts acting more like `break`. If you add `CmdletBinding` it'll work again.

There are other scenarios where that occurs but I can never remember them, it's come up before as a possible solution.

In order to get it to work in all scenarios, [this is what I do in my profile](<https://github.com/SeeminglyScience/dotfiles/blob/f71ae620d7b30246c33dec8636a4a609c449d517/Documents/PowerShell/Utility.psm1#L180-L257>) (though in case it needs to be said, this is very outside of support).

**Comment by TobiasPSP** [comment]
Thank you for your comment. Your code is raising the magic exception directly whereas the proposed solution from above indirectly does the same, just by using Select-Object as proxy.

Both ways work reliably. The reason why Select-Object failed in simple functions and worked in advanced functions is probably due to the way how proxy functions work. Occasionally users repurposed autogenerated sample code which may require $PSCmdlet (NULL in simple functions). The suggested solution above does not have this limitation.

That said, at this time the exception you raise directly is marked non-public, and although unlikely, it may change at any time without notice. So at this point, strictly speaking, only the Select-Object-approach is safe to use.

However, this approach is also a bit overkill. I would propose to properly document the exception and make it public, then either use this alone (raising the exception does the trick and is a one-liner), or add a simple wrapper cmdlet or keyword for discoverability.

While at this, maybe someone can look into the break keyword and (a) fix its runaway tendency when used outside loops that crosses even module boundaries and leads to very hard-to-identify cross-module issues currently, and then (b) add the pipeline break functionality to this existing keyword, based on the exception, again including proper safe-guarding (as this exception will also runaway when issued outside a pipeline). 

Making  „break“ also work properly in pipelines seems the most intuitive solution even though potential breaking changes would require scrutiny here.

**Comment by SeeminglyScience** [comment]
> Both ways work reliably.

Yeah so I don't remember the specifics, but the proxy does not always work. It's come up a few times in the discord as a potential solution until folks hit the blocker sometime after. If I can find it I'll post it here, but for whatever reason I always forget the exact issue that crops up.

> That said, at this time the exception you raise directly is marked non-public, and although unlikely, it may change at any time without notice.

Correct! Though both workarounds arguably rely on undocumented implementation detail. I don't suggest the use of either in production code.

> However, this approach is also a bit overkill. I would propose to properly document the exception and make it public, then either use this alone (raising the exception does the trick and is a one-liner), or add a simple wrapper cmdlet or keyword for discoverability.

Yeah, there are a few proposals out there that need to be discussed. I'm partial to adding a `PSCmdlet.StopUpstreamCommands()` method personally.

> While at this, maybe someone can look into the break keyword and (a) fix its runaway tendency when used outside loops that crosses even module boundaries and leads to very hard-to-identify cross-module issues currently, and then (b) add the pipeline break functionality to this existing keyword, based on the exception, again including proper safe-guarding (as this exception will also runaway when issued outside a pipeline).

That one, unfortunately, is by design behavior that is depended on by folks. For example:

```powershell
$b = 0..10
$stuffToSkip = 0, 2, 4, 6, 8, 10
foreach ($a in $b) {
    $stuffToSkip | % { if ($a -eq $_) { continue } }
    $a
}
```

I would have made it a parse error personally, but it is what it is.
